src/sql/src/query_model/mod.rs

```diff
@@ -0,0 +1,403 @@
+    /// guaranteed by structure of the box or it must preserve duplicated
+    /// rows from its input boxes. See [DistinctOperation].
+    pub distinct: DistinctOperation,
+}
+
+impl QueryBox {
+    /// Add all columns from the non-subquery input quantifiers of the box to the
+    /// projection of the box.
+    fn add_all_input_columns(&mut self, model: &Model) {
+        for quantifier_id in self.quantifiers.iter() {
+            let q = model.get_quantifier(*quantifier_id);
+            let bq = q.borrow();
+            if !bq.quantifier_type.is_subquery() {
+                let input_box = model.get_box(bq.input_box).borrow();
+                for (position, c) in input_box.columns.iter().enumerate() {
+                    let expr = Expr::ColumnReference(ColumnReference {
+                        quantifier_id: *quantifier_id,
+                        position,
+                    });
+                    self.columns.push(Column {
+                        expr,
+                        alias: c.alias.clone(),
+                    });
+                }
+            }
+        }
+    }
+
+    /// Visit all the expressions in this query box.
+    fn visit_expressions<F, E>(&self, f: &mut F) -> Result<(), E>
+    where
+        F: FnMut(&Expr) -> Result<(), E>,
+    {
+        for c in self.columns.iter() {
+            f(&c.expr)?;
+        }
+        match &self.box_type {
+            BoxType::Select(select) => {
+                for p in select.predicates.iter() {
+                    f(p)?;
+                }
+                if let Some(order_key) = &select.order_key {
+                    for p in order_key.iter() {
+                        f(p)?;
+                    }
+                }
+                if let Some(limit) = &select.limit {
+                    f(limit)?;
+                }
+                if let Some(offset) = &select.offset {
+                    f(offset)?;
+                }
+            }
+            BoxType::OuterJoin(outer_join) => {
+                for p in outer_join.predicates.iter() {
+                    f(p)?;
+                }
+            }
+            BoxType::Grouping(grouping) => {
+                for p in grouping.key.iter() {
+                    f(p)?;
+                }
+            }
+            BoxType::Values(values) => {
+                for row in values.rows.iter() {
+                    for value in row.iter() {
+                        f(value)?;
+                    }
+                }
+            }
+            BoxType::TableFunction(table_function) => {
+                for p in table_function.parameters.iter() {
+                    f(p)?;
+                }
+            }
+            BoxType::Except | BoxType::Union | BoxType::Intersect | BoxType::BaseTable(_) => {}
+        }
+        Ok(())
+    }
+
+    fn is_select(&self) -> bool {
+        matches!(self.box_type, BoxType::Select(_))
+    }
+}
+
+#[derive(Debug)]
+pub enum BoxType {
```

Comment:
Please also add a comment what each `BoxType` and `QuantifierType` variant represents.